mmedit/models/video_interpolater/basic_interpolater.py

```diff
@@ -0,0 +1,289 @@
+    """Basic model for video interpolation.
+
+    It must contain a generator that takes frames as inputs and outputs an
+    interpolated frame. It also has a pixel-wise loss for training.
+
+    The subclasses should overwrite the function `forward_train`,
+    `forward_test` and `train_step`.
+
+    Args:
+        generator (dict): Config for the generator structure.
+        pixel_loss (dict): Config for pixel-wise loss.
+        train_cfg (dict): Config for training. Default: None.
+        test_cfg (dict): Config for testing. Default: None.
+        pretrained (str): Path for pretrained model. Default: None.
+    """
+    allowed_metrics = {'PSNR': psnr, 'SSIM': ssim}
+
+    def __init__(self,
+                 generator,
+                 pixel_loss,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super().__init__()
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        # support fp16
+        self.fp16_enabled = False
+
+        # generator
+        self.generator = build_backbone(generator)
+        self.init_weights(pretrained)
+
+        # loss
+        self.pixel_loss = build_loss(pixel_loss)
+
+    def init_weights(self, pretrained=None):
+        """Init weights for models.
+
+        Args:
+            pretrained (str, optional): Path for pretrained weights. If given
+                None, pretrained weights will not be loaded. Defaults to None.
+        """
+        self.generator.init_weights(pretrained)
+
+    @auto_fp16(apply_to=('inputs', ))
+    def forward(self, inputs, target=None, test_mode=False, **kwargs):
+        """Forward function.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frames.
+            target (Tensor): Tensor of target frame. Default: None.
+            test_mode (bool): Whether in test mode or not. Default: False.
+            kwargs (dict): Other arguments.
+        """
+
+        if test_mode:
+            return self.forward_test(inputs, target, **kwargs)
+
+        return self.forward_train(inputs, target)
+
+    def forward_train(self, inputs, target):
+        """Training forward function.
+
+        This is a basic function, interpolate a frame between the given two
+        frames.
+
+        Args:
+            inputs (Tensor): Tensor of inputs frames with shape
+                (n, 2, c, h, w).
+            target (Tensor): Tensor of target frame with shape (n, c, h, w).
+
+        Returns:
+            Tensor: Output tensor.
+        """
+        losses = dict()
+        output = self.generator(inputs)
+        loss_pix = self.pixel_loss(output, target)
+        losses['loss_pix'] = loss_pix
+        outputs = dict(
+            losses=losses,
+            num_samples=len(target.data),
+            results=dict(
+                inputs=inputs.cpu(), target=target.cpu(), output=output.cpu()))
+        return outputs
+
+    def evaluate(self, output, target):
```

Comment:
If this function is for a model that produces outputs with `(n, c, h, w)`, we do not need to consider the case of  `(n, t, c, h, w)`.

Reply:
The output of CAIN is an image, but maybe some methods produce more than one image.